silkworm/rpc/http/connection.cpp

```diff
@@ -253,19 +291,50 @@ Connection::AuthorizationResult Connection::is_request_authorized(const boost::b
 }
 
 template <class Body>
-void Connection::set_cors(boost::beast::http::response<Body>& res) {
-    if (allowed_origins_.empty()) {
+void Connection::set_cors(const boost::beast::http::request<boost::beast::http::string_body>& req, boost::beast::http::response<Body>& res) {
+    bool is_preflight = (req.method() == boost::beast::http::verb::options && !req[boost::beast::http::field::access_control_request_method].empty());
+    std::string vary = req[boost::beast::http::field::vary];
+
+    if (is_preflight) {
+        if (vary.empty()) {
+            res.set(boost::beast::http::field::vary, "Origin, Access-Control-Request-Method, Access-Control-Request-Headers");
+        } else {
+            vary.append(" Origin");
+            res.set(boost::beast::http::field::vary, vary);
+        }
+    } else {
+        if (vary.empty()) {
+            res.set(boost::beast::http::field::vary, "Origin");
+        } else {
+            vary.append(" Origin");
+            res.set(boost::beast::http::field::vary, vary);
+        }
+    }
+
+    std::string origin = req[boost::beast::http::field::origin];
+    if (origin.empty()) {
+        return;
+    }
+
+    if (!is_origin_allowed(allowed_origins_, origin)) {
+        return;
+    }
+
+    if (!is_method_allowed(req.method())) {
         return;
     }
 
     if (allowed_origins_.at(0) == "*") {
-        res.set("Access-Control-Allow-Origin", "*");
+        res.set(boost::beast::http::field::access_control_allow_origin, "*");
     } else {
-        res.set("Access-Control-Allow-Origin", absl::StrJoin(allowed_origins_, ","));
+        res.set(boost::beast::http::field::access_control_allow_origin, origin);
+    }
+
+    if (is_preflight) {
+        res.set(boost::beast::http::field::access_control_request_method, req[boost::beast::http::field::access_control_request_method]);
+        res.set(boost::beast::http::field::access_control_allow_headers, allow_headers);
+        res.set(boost::beast::http::field::access_control_max_age, max_age);
     }
-    res.set("Access-Control-Allow-Methods", "GET, POST");
-    res.set("Access-Control-Allow-Headers", "*");
-    res.set("Access-Control-Max-Age", "600");
 }
```

Comment:
This `set_cors` method seems to be not so simple and clean, maybe because it's handling both the HTTP pre-flight and the actual HTTP request. You can try to improve code readability by splitting it and put the code relevant to pre-flight in `handle_preflight`